### PR TITLE
implementation fastEQCompareOperator

### PR DESCRIPTION
**Pull request by watason**
# Summary
implementation new fast EQoperator.
new EQ is faster than old.
the speed of old op is 120 millisecond per cycle, new op is 45 millisecond per cycle, about three times faster.
Other speed ups, such as asyncronize are also possible.
but it is another task.
this PR does not implement it.
# Purpose

# Contents

# Testing Methods Performed

### CODE REVIEW

**Comment by watason** [comment]
if review is complete, rebase the PR into single commit.

File: src/ComputationContainer/Share/Compare.hpp [hunk 1]
```diff
@@ -109,7 +110,140 @@ bool operator<(const Share<FixedPoint> &left, const FixedPoint &right);
 bool operator==(const Share<FixedPoint> &left, const FixedPoint &right);
 
 Share<FixedPoint> LTZ(const Share<FixedPoint> &s);
+/// @brief l=32 split array
+inline constexpr int delta[] = {5, 5, 5, 5, 5, 5, 2};
+/// @brief generate one-hot-vector used by unitv
+/// @tparam N
+/// @return random share [r] and r-th value is 1 ,otherwise is 0 in n length array
```
Comment (arukuka) [inline src/ComputationContainer/Share/Compare.hpp:117]: This docs seems to be for `unitvPrep`

File: src/ComputationContainer/Share/Compare.hpp [hunk 2]
```diff
@@ -109,7 +110,140 @@ bool operator<(const Share<FixedPoint> &left, const FixedPoint &right);
...
+template <size_t N>
+std::pair<Share<int>, std::vector<Share<int>>> unitvPrep()
+{
+    std::vector<Share<int>> e(N * N);
+    std::vector<AddressId> addressIds(N * N);
+    std::vector<Share<int>> ret(N * N);
+    Config *conf = Config::getInstance();
+    auto server = ComputationToComputation::Server::getServer();
+    auto random_s = RandGenerator::getInstance()->getRandVec<long long>(1, 1 << 20, N);
+    auto r = RandGenerator::getInstance()->getRand<long long>(0, 31);
```
Comment (arukuka) [inline src/ComputationContainer/Share/Compare.hpp:132]: is `r` related to `N`?
```suggestion
    auto r = RandGenerator::getInstance()->getRand<long long>(0, N - 1);
```

**Review by arukuka — CHANGES_REQUESTED** [review]
@watason Thank you for your contribution!  
I left a comment because I had some concerns.  
I would appreciate it if you could confirm it.

We would also appreciate it if you could provide a citation to the reference somewhere.

File: src/ComputationContainer/Share/Compare.hpp (hunk 1 above)
Comment (watason) [reply src/ComputationContainer/Share/Compare.hpp:117]: you are right!
i got it.

Comment (arukuka) [reply src/ComputationContainer/Share/Compare.hpp]: I checked it, thank you.

File: src/ComputationContainer/Share/Compare.hpp (hunk 2 above)
Comment (arukuka) [reply src/ComputationContainer/Share/Compare.hpp]: I checked it, thank you.

**Review by arukuka — APPROVED** [review]

File: src/ComputationContainer/Test/IntegrationTest/ShareTest.hpp [hunk 3]
```diff
@@ -1247,4 +1247,106 @@ TEST(ShareTest, FPMulExtraLarge)
...
+}
+
+TEST(ShareTest, expandTest)
+{
+    int x = 5;  // 00000011
+    auto d = qmpc::Share::expand(x);
+    for (auto a : d)
+    {
+        std::cout << a << std::endl;
+    }
```
Comment (mdonaka) [inline src/ComputationContainer/Test/IntegrationTest/ShareTest.hpp:1282]: I would like to see it replaced with a comparative test using EXPECT_EQ.

File: src/ComputationContainer/Random/Random.hpp [hunk 4]
```diff
@@ -28,5 +28,16 @@ class RandGenerator
     template <class T>
     std::vector<T> getRandVec(
         long long min_val = 0, long long max_val = 9223372036854775807ll, int n = 5
-    );
```
Comment (mdonaka) [inline src/ComputationContainer/Random/Random.hpp:31]: Please delete getRandVec on Random.cpp.
https://github.com/acompany-develop/QuickMPC/blob/6e72c1d5d792f8678690d78a3867ad2a2e3fa4ca/src/ComputationContainer/Random/Random.cpp#L39-L53

File: src/ComputationContainer/Test/IntegrationTest/ShareTest.hpp (hunk 3 above)
Comment (watason) [reply src/ComputationContainer/Test/IntegrationTest/ShareTest.hpp:1282]: ah sure!.
i got it.

**Review by mdonaka — APPROVED** [review]
I have confirmed the change. Thanks!

@watason 
> if review is complete, rebase the PR into single commit.

I am willing to leave it as is, do you want to rebase?

**Comment by watason** [comment]
@mdonaka 
Yes i really wanted to do that, but i merged develop by misstake.
 i should have a large number of change commit.
so i am not going to rebase.
sorry.

please merge it.

**Pull request merged**